tests/components/modbus/test_modbus_sensor.py

```diff
@@ -0,0 +1,355 @@
+"""The tests for the Modbus sensor component."""
+import pytest
+from datetime import timedelta
+from unittest import mock
+
+from homeassistant.const import (
+    CONF_NAME,
+    CONF_OFFSET,
+    CONF_PLATFORM,
+    CONF_SCAN_INTERVAL,
+)
+from homeassistant.components.modbus import DEFAULT_HUB, DOMAIN as MODBUS_DOMAIN
+from homeassistant.components.modbus.sensor import (
+    CONF_COUNT,
+    CONF_DATA_TYPE,
+    CONF_PRECISION,
+    CONF_REGISTER,
+    CONF_REGISTER_TYPE,
+    CONF_REGISTERS,
+    CONF_REVERSE_ORDER,
+    CONF_SCALE,
+    DATA_TYPE_FLOAT,
+    DATA_TYPE_INT,
+    DATA_TYPE_UINT,
+    REGISTER_TYPE_HOLDING,
+    REGISTER_TYPE_INPUT,
+)
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.setup import async_setup_component
+import homeassistant.util.dt as dt_util
+from tests.common import MockModule, mock_integration, async_fire_time_changed
+
+
+@pytest.fixture()
+def mock_hub(hass):
+    """Mock hub."""
+    mock_integration(hass, MockModule(MODBUS_DOMAIN))
+    hub = mock.MagicMock()
+    hub.name = "hub"
+    hass.data[MODBUS_DOMAIN] = {DEFAULT_HUB: hub}
+    return hub
+
+
+common_register_config = {CONF_NAME: "test-config", CONF_REGISTER: 1234}
+
+
+async def run_test(hass, mock_hub, register_config, register_words, expected):
+    """Run test for given config and check that sensor outputs expected result."""
+
+    # Full sensor configuration
+    sensor_name = "modbus_test_sensor"
+    scan_interval = 5
+    config = {
+        SENSOR_DOMAIN: {
+            CONF_PLATFORM: "modbus",
+            CONF_SCAN_INTERVAL: scan_interval,
+            CONF_REGISTERS: [
+                dict(**{CONF_NAME: sensor_name, CONF_REGISTER: 1234}, **register_config)
+            ],
+        }
+    }
+
+    class ReadResult:
+        def __init__(self, register_words):
+            self.registers = register_words
+
+    # Setup inputs for the sensor
+    read_result = ReadResult(register_words)
+    if register_config.get(CONF_REGISTER_TYPE) == REGISTER_TYPE_INPUT:
+        mock_hub.read_input_registers.return_value = read_result
+    else:
+        mock_hub.read_holding_registers.return_value = read_result
+
+    # Initialize sensor
+    assert await async_setup_component(hass, SENSOR_DOMAIN, config)
+
+    # Trigger update call with time_changed event
+    future = dt_util.utcnow() + timedelta(seconds=scan_interval + 1)
```

Comment:
We need to patch time, in the correct place, to return the correct time, before setting up the component.

I think we should patch here:
https://github.com/home-assistant/home-assistant/blob/73f55757088709214f3498b56ebb8780dc5e8217/homeassistant/helpers/event.py#L20

Reply:
Ok